### PR TITLE
:sparkles: Add Ignition support

### LINKED ISSUE #1006: Support CoreOS Ignition for Flatcar Linux
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
Tracking issue for FlatCar Linux support

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api-provider-vsphere version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`): 

cc @dongsupark

### PR DESCRIPTION
**Pull request by johananl**
**What this PR does / why we need it**:

This PR adds Ignition support to CAPV and implements the first workload cluster flavor which uses Ignition based on Flatcar Container Linux.

Supersedes #1093.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #1006

**Special notes for your reviewer**:

TODO:

- [x] Update to latest API version?
- [x] Add Ignition flavor?
- [x] Add/update e2e tests
- [x] Figure out how to specify Flatcar image version in CI
- [x] Push Flatcar e2e image to test infra
- [x] Once https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/1618 is merged, run `make generate` and commit the new generated template
- [ ] ~Add a feature gate?~
- [x] Document usage
- [x] Remove TMP commits
- [x] Squash commits

**Release note**:

```release-note
Add support for Ignition bootstrap and add a Flatcar-based cluster flavor.
```

### CODE REVIEW

**Comment by srm09** [comment]
I will take sometime to go through the entire PR, but for starters, this will require an explicit E2E test which should be run periodically, and we can talk about to make it PR Blocking or not.

**Comment by johananl** [comment]
> I will take sometime to go through the entire PR, but for starters, this will require an explicit E2E test which should be run periodically, and we can talk about to make it PR Blocking or not.

Yes, definitely. There is a TODO for adding tests in the description.

File: test/e2e/data/infrastructure-vsphere/kustomization/ignition/ignition.yaml
Comment (johananl) [inline test/e2e/data/infrastructure-vsphere/kustomization/ignition/ignition.yaml:240]: TODO: Remove the hardcoding. I need to figure out where to publish the image, which version to use and how to reference it properly in the e2e config.

**Comment by geetikabatra** [comment]
@johananl Could you also add a readme about how to utilize ignition support.

**Comment by johananl** [comment]
> @johananl Could you also add a readme about how to utilize ignition support. Reference https://cluster-api.sigs.k8s.io/tasks/experimental-features/ignition.html

Sure. By the way, do we have a docs website for CAPV?

**Comment by geetikabatra** [comment]
Not yet. I am filing an issue for the same.

**Comment by johananl** [comment]
Current status: waiting for https://github.com/kubernetes-sigs/image-builder/pull/817 (which is in turn blocked by https://github.com/kubernetes-sigs/image-builder/pull/827) so that we can execute the e2e tests I added which require Flatcar VMs.

**Comment by johananl** [comment]
@alexander-demichev I assume you're busy and don't want to bug you. For the sake of driving this PR forward, I'm going to rewrite the commit to fix the author problem and put you under `Co-authored-by` in the hopes that doing so would fix the problem. In case you have an up-to-date email for me, please talk to me soon :pray:

**Comment by johananl** [comment]
Thanks for providing the new address @alexander-demichev. I've amended the commit and CLA looks good now.

**Comment by johananl** [comment]
I have the following questions to the maintainers:

1. Do we need a feature gate for Ignition support? Ignition support is [gated](https://github.com/kubernetes-sigs/cluster-api/blob/daab6fdc72704b39e160878acf743c333648ddbb/feature/feature.go#L57) in core CAPI, however in CAPV I don't see any gates implemented at the moment.
2. `pull-cluster-api-provider-vsphere-apidiff-main` is failing because I've modified a couple of function signatures in `pkg/services/govmomi`. This package is an implementation detail AFAICT, so I'm not sure why we need to be backward compatible inside of it. Anyway, in case I'm missing something and we do - what is the process for making the necessary changes there?
3. With Ignition e2e tests now in place, should we make them PR-Blocking? IMO we want to ensure we aren't breaking CAPV for Ignition-powered distros when making changes so I'd say yes, however I wanted to hear your thoughts since doing so will add a bit of CI time.

Thanks!

**Comment by johananl** [comment]
For the sake of making progress, I'll make some assumptions regarding the answers to my questions above. When we get eyes on this again I'll correct anything I've assumed wrong.

- Reading through other PRs in this repo, it seems the apidiff job isn't required for merging and rather serves as a warning. Let's ignore it then since this PR can't be merged without API changes since we have to start supporting bootstrap format in `pkg/services/govmomi`.
- I'm going to assume we don't need a feature gate.
- I'm going to assume it's a good idea to make the Ignition e2e test PR-blocking to ensure Ignition support doesn't break.

**Comment by johananl** [comment]
Ready for merging as far as I'm concerned @MaxRink @vrabbi @srm09. I can do a final squash once reviewed.

**Comment by johananl** [comment]
Ready for merging IMO.

**Comment by srm09** [comment]
Could you squash the commits and I will make a final pass at it in the meantime. We can get this merged then. Sorry for the very long wait on this one.

File: test/e2e/ignition_test.go
Comment (srm09) [inline test/e2e/ignition_test.go:29]:
```suggestion
	capi_e2e.QuickStartSpec(ctx, func() capi_e2e.QuickStartSpecInput {
```
Use the context that is used in all e2e tests.

**Review by srm09 — CHANGES_REQUESTED** [review]
few nits, but overall lgtm. Looking to get this merged once the comments get resolved.

File: packaging/flavorgen/flavors/flavors.go
Comment (johananl) [reply packaging/flavorgen/flavors/flavors.go:144]: Yes, I'm not sure why this parameter is even necessary. I admit I've copy-pasted it from `controlPlane := newKubeadmControlplane(444, cpMachineTemplate, nil)`. Let me remove it.

Comment (johananl) [reply packaging/flavorgen/flavors/flavors.go:144]: FWIW I find the "JIT" replacement unexpected and quite confusing, but I guess changing this is out of scope for this PR.

File: .gitignore
Comment (johananl) [reply .gitignore]: Done.

File: packaging/flavorgen/flavors/flavors.go
Comment (johananl) [reply packaging/flavorgen/flavors/flavors.go]: Fixed. I've also removed another occurrence in `newKubeadmControlplane()` in a separate commit.

**Comment by johananl** [comment]
@srm09 all feedback is handled and the commits are squashed. CI is green except `pull-cluster-api-provider-vsphere-apidiff-main` but IIRC this job is more of an FYI than a blocker (correct?).

Thanks for consistently being a great maintainer! 🙏

**Comment by johananl** [comment]
Force-pushed again to fix a problem in the test case description.

**Pull request merged**
